packages/core/src/utils/htmlValidationUtil.ts

```diff
@@ -0,0 +1,73 @@
+// eslint-disable-next-line max-classes-per-file
+import * as cheerio from 'cheerio';
+import * as logger from './logger';
+
+const rules: Rule[] = [];
+
+export function validateHtmlWithRules(content: string, path: string): void {
+  const $ = cheerio.load(content);
+  rules.forEach((rule) => {
+    rule.validateAndLog($, path);
+  });
+}
+
+class Rule {
+  description: string = '';
+  causeHydration: boolean = false;
+  // eslint-disable-next-line lodash/prefer-constant
+  isValidDom: (rootNode: cheerio.Root) => boolean = () => true;
+  validateAndLog(rootNode: cheerio.Root, path: string) {
+    if (!this.isValidDom(rootNode)) {
+      if (this.causeHydration) {
+        logger.error(`Invalid HTML in ${path}. ${this.description}`);
+      } else {
+        logger.warn(`Invalid HTML in ${path}. ${this.description}`);
```

Comment:
Maybe you can differentiate the error and warning messages to make it more clear. The error message can be something more serious like `Failed to process invalid HTML in ${path}, ${description}` while the warning message can be something more informative like `Invalid HTML detected in ${path}, ${decription}` similar to what you have now

Reply:
Resolve as no longer relevant